Adding hirte integration test

Yarboa opened the pull request:
Resolve #118 
Adding hirte integration test
ubuntu support podman v3.44

hirte running as container with systemd installed
hirte containers should run as non root

Adding tests directory
Container files will be hosted here
Later on adding the test itself
run nodes ans orch in background

Currently testing functionality of node registration
Updated with latest changes

Yarboa committed:
Adding hirte integration test
ubuntu support podman v3.44

hirte running as container with systemd installed
hirte-orch should run as non root

Adding tests directory
Container files will be hosted here
Later on adding the test itself
run nodes ans orch in background

Currently testig functionality of node registration
Updated with latest changes

Signed-off-by: Yariv Rachmani <yrachman@redhat.com>

Yarboa commented on .github/workflows/e2e.yaml:
```diff
@@ -0,0 +1,64 @@
+  test:
+    name: cs9 x86_64 build
+    runs-on: ubuntu-latest
+    container:
```
IT is already running on container, please change accordingly

alexlarsson commented on .github/workflows/e2e.yaml:
```diff
@@ -0,0 +1,64 @@
+
+      - name: run hirte containers
+        run: |
+          podman run -d --rm --net host --systemd=always --name hirte-mgr -it \
```
It feels a bit weird to run in containers and the use --net host. Can't you just expose the right ports between the containers?

Yarboa replied:
sure

Yarboa reviewed (commented):
@sdunnagan @alexlarsson 

I do not know which test framework to use related content assertions for tests
pytest? meson test? go? please suggest

rhatdan commented on tests/Containerfile-build:
```diff
@@ -0,0 +1,11 @@
+
+ARG USERNAME=test-user
+
+RUN dnf install -y dnf-plugin-config-manager
```
Can you combine all of your run lines onto a single line, to prevent propagation of layers, also add a dnf -y clean all to the last line to remove execess noice in the image.  This will help shrink the size of the image.

Yarboa replied:
Thanks, forgot about the the layers, Ack
And clean saved tens of Megas

ygalblum commented on doc/test/hirte_e2e.md:
```diff
@@ -0,0 +1,18 @@
+- Containerfile
+
+Containerfile-build published in quay
+quay.io/yarboa/hirte-builder
```
Maybe this can be done in a later step. But, I think we should move this image to the containers organization in quay: https://quay.io/organization/containers

Yarboa replied:
Ack, will need your help

ygalblum replied:
I don't have the permissions for this organization. @rhatdan can you help?

rhatdan replied:
I don't have access to that (I think)
@edsantiago @TomSweeneyRedHat PTAL

edsantiago replied:
`organization/containers` has three images AFAICS, and those are the autobuild podman-buildah-skopeo containers. I don't think we want to push internal images there. I think you want `quay.io/libpod`.

ygalblum replied:
@edsantiago sure. But, we don't have access to the `libpod` organization either. How should it be set up?

mwperina replied:
We are not going to distribute hirte as a containers, it doesn't make sense. And here we are creating containers for each test run, so I don't see a reason why such containers should be published on quay.io. Or am I missing something?

Yarboa replied:
I will explain,
Currently the build workflow download the following:
https://github.com/containers/hirte/blob/main/.github/workflows/build.yml#L14-L29

Is running on css9 container, refer,  https://github.com/containers/hirte/blob/main/.github/workflows/build.yml#L13-L14
It install rpms, 
this container quay.io/libpod/hirte-builder or other name, replace L14-L29

Each time build run it pulls the build image and starts with meson commands

@ygalblum suggest to move this part into the build workflow, prepare test containers, and in e2e workflow consume hirte-images for e2e tests,

Does it make sense?

ygalblum replied:
The containers created each run are created for testing purposes and are not published at all. To reduce their build time, @Yarboa created a base container , `hirte-builder` (which may change, but only when needed), currently published in his private user account in Quay. 

Keeping containers images on someone's private account is a bad practice as it means that only they can push new images. For that I proposed that we find a better place to store this container. I suggested quay.io, but maybe a better place to put it would be here in GitHub along side the code.

Yarboa replied:
@ygalblum 
Issue created #140

Yarboa replied:
> @edsantiago sure. But, we don't have access to the `libpod` organization either. How should it be set up?

@edsantiago  @TomSweeneyRedHat Any Ideas are welcome
I am trying also test PR upload to ghcr with docker no success #144

Yarboa replied:
@pypingou @ygalblum 
Is there any reason not pusing into AutoSD repo? 
https://gitlab.com/CentOS/automotive/sample-images/-/blob/main/sample-apps/Makefile.container#L1

Yarboa replied:
Uploading to registry
registry.gitlab.com/centos/automotive/sample-images/hirte/hirte-builder:1.0.0

Yarboa replied:
Found temporary home, till  decision #140 is made

ygalblum commented on tests/Containerfile-build:
```diff
@@ -0,0 +1,13 @@
+RUN dnf config-manager -y --set-enabled crb && \
+    dnf install -y clang-tools-extra gcc make \
+    meson systemd-devel systemd
+RUN dnf -y clean all
```
I think it's better to run `dnf clean all` in the same layer that you run the install. AFAIK in the current way, you're not saving any storage

Yarboa replied:
Ack,
Note: it went from 1.37G -> 1.01G

ygalblum commented on .github/workflows/e2e.yaml:
```diff
@@ -0,0 +1,69 @@
+
+      - name: build hirte tests containers
+        run: |
+          podman run --rm --systemd=always --name hirte-builder \
```
I think these steps should go into the Makefile as targets and then the GitHub workflow will only trigger them. This way you allow the developer to run these tests locally

Yarboa replied:
Thanks, I will contact to understand it better

Yarboa replied:
@ygalblum
Issue created https://github.com/containers/hirte/issues/140

ygalblum commented on .github/workflows/e2e.yaml:
```diff
@@ -0,0 +1,69 @@
+
+      - name: run hirte orchestrator
+        run: |
+          podman exec -t  $(podman ps | grep hirte-mgr | awk '{print $1}') \
```
Since you set the name of the container, do you really need to look for it in `podman ps`? (same for all cases)

Yarboa replied:
Oops. you are correct

rhatdan replied:
In some cases you can take advantage of `podman exec -t -l ...`  to exec into the last started container, as long as you are executing commands locally, not on a Mac or Windows platform.

rhatdan commented on .github/workflows/e2e.yaml:
```diff
@@ -0,0 +1,69 @@
+
+      - name: run hirte containers
+        run: |
+          podman run -d --rm -p 2020:2020 --systemd=always --name hirte-mgr \
```
Why are you running with --systemd=always?  If your container defaults to running systemd, then Podman will pick this up.  --systemd=always is usually just used for debugging what is going on in a systemd environment and should not be used in general.

Yarboa replied:
Ack

Yarboa committed:
Adding generic repo

Till issue #140 agreed, using
registry.gitlab.com/centos/automotive/sample-images
To pull the image only for e2e testing

Signed-off-by: Yariv Rachmani <yrachman@redhat.com>

ygalblum commented on doc/test/hirte_e2e.md:
```diff
@@ -0,0 +1,24 @@
+## Environment Setup
+
+E2E tests rely on containers as seprate compute entities,
+Use of containers enble to simulaute hirte functional behaviour
```
fix `enble` -> `enable`

Yarboa replied:
Thanks, oh my :)
found this also 
seprate -> separate

Yarboa replied:
Done

Yarboa committed:
update hirte_e2e.md

Signed-off-by: Yariv Rachmani <yrachman@redhat.com>

engelmi reviewed (approved):
LGTM

Question: I am also able to run the e2e tests locally, right? I think it would be great to provide a script or Makefile target to simplify this. This would be awesome to implement in a follow-up PR.
